### ISSUE TITLE
ManyToOne with JoinColumn(nullable=false) doctrine.associationType false positive

### ISSUE THREAD

**Issue by mesolaries** [phpstan/phpstan-doctrine#766]
Consider the following association:

```php
class Foo
{
    /**
     * @ORM\ManyToOne(targetEntity=Bar::class, inversedBy="foos")
     * @ORM\JoinColumn(nullable=false)
     */
    private ?Bar $bar = null;

    public function setBar(?Bar $bar): self
    {
        $this->bar = $bar;
        
        return $this;
    }
}
```

PHPStan reports:
`Property Foo::$bar type mapping mismatch: property can contain Bar|null but database expects Bar`

I believe this is a false positive.

In Doctrine entities, it is common for a ManyToOne association mapped with `JoinColumn(nullable=false)` to be nullable and initialized to `null` before the entity is fully constructed or persisted. Also, Symfony generated `remove* `methods for bidirectional associations not work because they set the owning side to `null`.

Expected: no doctrine.associationType error for this scenario.

**Comment by VincentLanglet** [comment]
Hi, this is not a false positive, given the fact you are allowing an invalid state for your data 

It's the same that having a nullable field "name" which is required in database. Even if you have an explanation for this it is still a technical inconsistency which is worth being reported.

I personnaly have no such thing you can either:
- pass the value directly in the constructor
- using an uninitialized property


Another example of the issue, you are allowing to set null in your setter and then persist the data, this will crash. This is a bug, phpstan is reporting it.
